### ISSUE TITLE
Saga's handling same timeoutmessage type both get invoked

### ISSUE THREAD

**Issue by Tsjunne** [Particular/NServiceBus#4585]
When requesting a timeout in a saga with a timeout state X, every saga that uses the same timeout state X gets invoked if they share the same SagaData type. 

Despite the message headers containing the correct saga type in 'NServicebus.SagaType', timeouts requested in one saga type get processed in other types.

**Comment by andreasohlund** [comment]
What version of NServiceBus core is this? What saga persister + version are you using?

**Comment by andreasohlund** [comment]
How does you sagas look like, I was trying to repro this with this setup:

```
            public class ReuseSaga1 : Saga<ReuseSagaData>, IAmStartedByMessages<StartSagaMessage>,IHandleTimeouts<ReusedTimeout>
            {
                public Context Context { get; set; }

                public Task Handle(StartSagaMessage message, IMessageHandlerContext context)
                {
                    return RequestTimeout<ReusedTimeout>(context,TimeSpan.FromMilliseconds(10));
                }

                protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ReuseSagaData> mapper)
                {
                    mapper.ConfigureMapping<StartSagaMessage>(m=>m.SomeId).ToSaga(s=>s.SomeId);
                }

                public Task Timeout(ReusedTimeout state, IMessageHandlerContext context)
                {
                    Context.TimeoutInvokedForSaga1 = true;
                    return Task.FromResult(0);
                }
            }

            public class ReuseSaga2 : Saga<ReuseSagaData>, IAmStartedByMessages<StartSagaMessage>, IHandleTimeouts<ReusedTimeout>
            {
                public Context Context { get; set; }

                public Task Handle(StartSagaMessage message, IMessageHandlerContext context)
                {
                    return Task.FromResult(0);
                }

                public Task Timeout(ReusedTimeout state, IMessageHandlerContext context)
                {
                    Context.TimeoutInvokedForSaga2 = true;

                    return Task.FromResult(0);
                }

                protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ReuseSagaData> mapper)
                {
                    mapper.ConfigureMapping<StartSagaMessage>(m => m.SomeId).ToSaga(s => s.SomeId);
                }

                public class ReuseSagaData1 : ContainSagaData
                {
                    public virtual string SomeId { get; set; }
                }
            }
            public class ReuseSagaData : ContainSagaData
            {
                public virtual string SomeId { get; set; }
            }
            public class ReusedTimeout
            {

            }
```

But that won't even run since the persister will try to create the same instance twice and hit a concurrency error?

**Comment by andreasohlund** [comment]
>The start message was different in our case. Only the timeout message was the same.

Does the start messages have the same correlation id? 

Ie is this test reproducing your scenario? https://github.com/Particular/NServiceBus/blob/saga-data-reuse/src/NServiceBus.AcceptanceTests/Core/When_reusing_saga_data.cs

**Comment by Tsjunne** [comment]
Ah yes, that would have been the case.

**Comment by andreasohlund** [comment]
In that case the inmemory persister prevents me from repro it so I'll move this to the nhibernate repo

Reusing saga timeouts and sagastate us usually a very bad idea, can you share the business scenario you are targeting so that we can try to find  a better design for you?

**Comment by Tsjunne** [comment]
I figured as much :) It was one of our devs that stumbled upon the issue. It was just a case of code reusability for a bunch of saga's that did pretty much the same thing, but different contexts. But we solved it since, using a discriminator for the sagadata.
So if this scenario isn't anything that should be supported, feel free to close the issue.

**Comment by andreasohlund** [comment]
@Particular/nhibernate-persistence-maintainers can you eyeball this test

https://github.com/Particular/NServiceBus/blob/saga-data-reuse/src/NServiceBus.AcceptanceTests/Core/When_reusing_saga_data.cs

would NHibernate create a table per saga even though the saga entity type is shared?

**Comment by MarcinHoppe** [comment]
@andreasohlund One table will be created per saga data type. In this case, there will be only one table.

**Comment by andreasohlund** [comment]
@Tsjunne that seems to explain the vierd behavior. We'll do the following:

* Prevent saga entity reuse as of 6.3 https://github.com/Particular/NServiceBus/issues/4597
* Try to improve the persister API to use saga types when naming storage entities like tables/folders/collections etc
* Figure out why the invocation happened even though the header was present to indicate that the timeout was supposed to hit only one of the sagas

**Comment by Tsjunne** [comment]
#4596 Looks like a pretty big change, that would not necessarily feel more intuitive, since data and behavior are already split up in the model and data is the thing that gets persisted. 
I guess #4597 could prevent this weird behavior from occuring.

The only thing that is truly wrong is the fact that a saga instance processes a timeout that was clearly created by another saga and labeled as such through the NServicebus.SagaType header. And i'm not entirely sure if preventing data reuse would solve this. It looks like this could happen when 2 saga's handle the same timeout message type and have the same ID coincidentally, wich could also happen with different sagadata as well. I would expect NSB to explicitly prevent saga's being loaded from another type than the one registered in the header.

PS: I checked https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs again and it looks like this was fixed in https://github.com/Particular/NServiceBus/commit/9352637b32c89ecde328f776ae7b93fc9cb366a5

**Comment by andreasohlund** [comment]
>I checked https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs again and it looks like this was fixed in 9352637

Ah gotcha, I though we had that code in 6.1 that's why I thought this was a new bug.

Would you have any chance to upgrade to 6.2 to confirm that this is fixed?

**Comment by Tsjunne** [comment]
Hmm, well we already created different sagadata types and made the old shared one abstract. This served as a workaround. But i checked the code that i pulled in a few weeks ago and I remarked that no check was done on the saga type and this would cause the behavior. The code that was missing is exactly what was added in 9352637. 
So as far as i am concerned, this is fixed in 6.2, and there is no need to act on any of the bulletpoints you listed for what this issue is concerned.

**Comment by andreasohlund** [comment]
>So as far as i am concerned, this is fixed in 6.2, and there is no need to act on any of the bulletpoints you listed for what this issue is concerned.

Awesome, still thing the saga name is a better and more intuitive name for tables etc. We'll discuss that internally and see what gives.

Thanks for all the help!
